File: ClassicalSharp/Math/Physics/Searcher.cs
// Copyright 2014-2017 ClassicalSharp | Licensed under BSD-3
using ClassicalSharp.Entities;
using ClassicalSharp.Map;
using System;
using OpenTK;

namespace ClassicalSharp.Physics {
	
	public struct State {
		public int X, Y, Z;
		public float tSquared;
		
		public State(int x, int y, int z, byte block, float tSquared) {
			X = x << 3; Y = y << 3; Z = z << 3;
			X |= (block & 0x07);
			Y |= (block & 0x38) >> 3;
			Z |= (block & 0xC0) >> 6;
			this.tSquared = tSquared;
		}
	}
	
	/// <summary> Calculates all possible blocks that a moving entity can intersect with. </summary>
	public sealed class Searcher {
		
		public static State[] stateCache = new State[0];
		
		public static int FindReachableBlocks(Game game, Entity entity,
		                                      out AABB entityBB, out AABB entityExtentBB) {
			Vector3 vel = entity.Velocity;
			entityBB = entity.Bounds;
			
			// Exact maximum extent the entity can reach, and the equivalent map coordinates.
			entityExtentBB = new AABB(
				vel.X < 0 ? entityBB.Min.X + vel.X : entityBB.Min.X,
				vel.Y < 0 ? entityBB.Min.Y + vel.Y : entityBB.Min.Y,
				vel.Z < 0 ? entityBB.Min.Z + vel.Z : entityBB.Min.Z,
				vel.X > 0 ? entityBB.Max.X + vel.X : entityBB.Max.X,
				vel.Y > 0 ? entityBB.Max.Y + vel.Y : entityBB.Max.Y,
				vel.Z > 0 ? entityBB.Max.Z + vel.Z : entityBB.Max.Z
			);
			Vector3I min = Vector3I.Floor(entityExtentBB.Min);
			Vector3I max = Vector3I.Floor(entityExtentBB.Max);
			
			int elements = (max.X + 1 - min.X) * (max.Y + 1 - min.Y) * (max.Z + 1 - min.Z);
			if (elements > stateCache.Length) {
				stateCache = new State[elements];
			}
			
			
			AABB blockBB = default(AABB);
			BlockInfo info = game.BlockInfo;
			int count = 0;
			// Order loops so that we minimise cache misses
			for (int y = min.Y; y <= max.Y; y++)
				for (int z = min.Z; z <= max.Z; z++)
					for (int x = min.X; x <= max.X; x++)
			{
				byte block = game.World.GetPhysicsBlock(x, y, z);
				if (info.Collide[block] != CollideType.Solid) continue;
				
				blockBB.Min = info.MinBB[block];
				blockBB.Min.X += x; blockBB.Min.Y += y; blockBB.Min.Z += z;
				blockBB.Max = info.MaxBB[block];
				blockBB.Max.X += x; blockBB.Max.Y += y; blockBB.Max.Z += z;
				
				if (!entityExtentBB.Intersects(blockBB)) continue; // necessary for non whole blocks. (slabs)
				
				float tx = 0, ty = 0, tz = 0;
				CalcTime(ref vel, ref entityBB, ref blockBB, out tx, out ty, out tz);
				if (tx > 1 || ty > 1 || tz > 1) continue;
				float tSquared = tx * tx + ty * ty + tz * tz;
				stateCache[count++] = new State(x, y, z, block, tSquared);
			}
			
			if (count > 0)
				QuickSort(stateCache, 0, count - 1);
			return count;
		}

		
		public static void CalcTime(ref Vector3 vel, ref AABB entityBB, ref AABB blockBB,
		                            out float tx, out float ty, out float tz) {
			float dx = vel.X > 0 ? blockBB.Min.X - entityBB.Max.X : entityBB.Min.X - blockBB.Max.X;
			float dy = vel.Y > 0 ? blockBB.Min.Y - entityBB.Max.Y : entityBB.Min.Y - blockBB.Max.Y;
			float dz = vel.Z > 0 ? blockBB.Min.Z - entityBB.Max.Z : entityBB.Min.Z - blockBB.Max.Z;
			
			tx = vel.X == 0 ? float.PositiveInfinity : Math.Abs(dx / vel.X);
			ty = vel.Y == 0 ? float.PositiveInfinity : Math.Abs(dy / vel.Y);
			tz = vel.Z == 0 ? float.PositiveInfinity : Math.Abs(dz / vel.Z);
			
			if (entityBB.Max.X >= blockBB.Min.X && entityBB.Min.X <= blockBB.Max.X) tx = 0; // Inlined XIntersects
			if (entityBB.Max.Y >= blockBB.Min.Y && entityBB.Min.Y <= blockBB.Max.Y) ty = 0; // Inlined YIntersects
			if (entityBB.Max.Z >= blockBB.Min.Z && entityBB.Min.Z <= blockBB.Max.Z) tz = 0; // Inlined ZIntersects
		}
		
		static void QuickSort(State[] keys, int left, int right) {
			while (left < right) {
				int i = left, j = right;
				float pivot = keys[(i + j) / 2].tSquared;
				// partition the list
				while (i <= j) {
					while (pivot > keys[i].tSquared) i++;
					while (pivot < keys[j].tSquared) j--;
					
					if (i <= j) {
						State key = keys[i]; keys[i] = keys[j]; keys[j] = key;
						i++; j--;
					}
				}
				
				// recurse into the smaller subset
				if (j - left <= right - i) {
					if (left < j)
						QuickSort(keys, left, j);
					left = i;
				} else {
					if (i < right)
						QuickSort(keys, i, right);
					right = j;
				}
			}
		}
	}
}

File: src/Client/Searcher.c
#include "Searcher.h"
#include "Block.h"
#include "World.h"
#include "Platform.h"
#include "ExtMath.h"
#include "Funcs.h"

void SearcherState_Init(SearcherState* state, Int32 x, Int32 y, Int32 z, BlockID block, Real32 tSquared) {
	state->X = (x << 3) | (block & 0x07);
	state->Y = (y << 3) | (block & 0x38) >> 3;
	state->Z = (z << 3) | (block & 0xC0) >> 6;
	state->tSquared = tSquared;
}

#define SEARCHER_STATES_MIN 64
SearcherState Searcher_StatesInitial[SEARCHER_STATES_MIN];
extern SearcherState* Searcher_States = Searcher_StatesInitial;
UInt32 Searcher_StatesCount = SEARCHER_STATES_MIN;

void Searcher_QuickSort(Int32 left, Int32 right) {
	SearcherState* keys = Searcher_States; SearcherState key;
	while (left < right) {
		Int32 i = left, j = right;
		Real32 pivot = keys[(i + j) >> 1].tSquared;

		/* partition the list */
		while (i <= j) {
			while (pivot > keys[i].tSquared) i++;
			while (pivot < keys[j].tSquared) j--;
			QuickSort_Swap_Maybe();
		}
		/* recurse into the smaller subset */
		QuickSort_Recurse(Searcher_QuickSort);
	}
}

UInt32 Searcher_FindReachableBlocks(Entity* entity, AABB* entityBB, AABB* entityExtentBB) {
	Vector3 vel = entity->Velocity;
	Entity_GetBounds(entity, entityBB);

	/* Exact maximum extent the entity can reach, and the equivalent map coordinates. */
	entityExtentBB->Min.X = entityBB->Min.X + (vel.X < 0.0f ? vel.X : 0.0f);
	entityExtentBB->Min.Y = entityBB->Min.Y + (vel.Y < 0.0f ? vel.Y : 0.0f);
	entityExtentBB->Min.Z = entityBB->Min.Z + (vel.Z < 0.0f ? vel.Z : 0.0f);

	entityExtentBB->Max.X = entityBB->Max.X + (vel.X > 0.0f ? vel.X : 0.0f);
	entityExtentBB->Max.Y = entityBB->Max.Y + (vel.Y > 0.0f ? vel.Y : 0.0f);
	entityExtentBB->Max.Z = entityBB->Max.Z + (vel.Z > 0.0f ? vel.Z : 0.0f);

	Vector3I min, max;
	Vector3I_Floor(&min, &entityExtentBB->Min);
	Vector3I_Floor(&max, &entityExtentBB->Max);

	UInt32 elements = (max.X - min.X + 1) * (max.Y - min.Y + 1) * (max.Z - min.Z + 1);
	if (elements > Searcher_StatesCount) {
		if (Searcher_StatesCount > SEARCHER_STATES_MIN) {
			Platform_MemFree(Searcher_States);
		}
		Searcher_StatesCount = elements;

		Searcher_States = Platform_MemAlloc(elements * sizeof(SearcherState));
		if (Searcher_States == NULL) {
			ErrorHandler_Fail("Failed to allocate memory for Searcher_FindReachableBlocks");
		}
	}

	/* Order loops so that we minimise cache misses */
	AABB blockBB;
	UInt32 count = 0;
	Int32 x, y, z;
	SearcherState* curState = Searcher_States;

	for (y = min.Y; y <= max.Y; y++) {
		for (z = min.Z; z <= max.Z; z++) {
			for (x = min.X; x <= max.X; x++) {
				BlockID block = World_GetPhysicsBlock(x, y, z);
				if (Block_Collide[block] != COLLIDE_SOLID) continue;
				Real32 xx = (Real32)x, yy = (Real32)y, zz = (Real32)z;

				blockBB.Min = Block_MinBB[block];
				blockBB.Min.X += xx; blockBB.Min.Y += yy; blockBB.Min.Z += zz;
				blockBB.Max = Block_MaxBB[block];
				blockBB.Max.X += xx; blockBB.Max.Y += yy; blockBB.Max.Z += zz;

				if (!AABB_Intersects(entityExtentBB, &blockBB)) continue; /* necessary for non whole blocks. (slabs) */

				Real32 tx, ty, tz;
				Searcher_CalcTime(&vel, entityBB, &blockBB, &tx, &ty, &tz);
				if (tx > 1.0f || ty > 1.0f || tz > 1.0f) continue;
				Real32 tSquared = tx * tx + ty * ty + tz * tz;

				SearcherState_Init(curState, x, y, z, block, tSquared);
				curState++;
			}
		}
	}

	if (count > 0) Searcher_QuickSort(0, count - 1);
	return count;
}


void Searcher_CalcTime(Vector3* vel, AABB *entityBB, AABB* blockBB, Real32* tx, Real32* ty, Real32* tz) {
	Real32 dx = vel->X > 0.0f ? blockBB->Min.X - entityBB->Max.X : entityBB->Min.X - blockBB->Max.X;
	Real32 dy = vel->Y > 0.0f ? blockBB->Min.Y - entityBB->Max.Y : entityBB->Min.Y - blockBB->Max.Y;
	Real32 dz = vel->Z > 0.0f ? blockBB->Min.Z - entityBB->Max.Z : entityBB->Min.Z - blockBB->Max.Z;

	*tx = vel->X == 0.0f ? MATH_POS_INF : Math_AbsF(dx / vel->X);
	*ty = vel->Y == 0.0f ? MATH_POS_INF : Math_AbsF(dy / vel->Y);
	*tz = vel->Z == 0.0f ? MATH_POS_INF : Math_AbsF(dz / vel->Z);

	if (entityBB->Max.X >= blockBB->Min.X && entityBB->Min.X <= blockBB->Max.X) *tx = 0.0f; /* Inlined XIntersects */
	if (entityBB->Max.Y >= blockBB->Min.Y && entityBB->Min.Y <= blockBB->Max.Y) *ty = 0.0f; /* Inlined YIntersects */
	if (entityBB->Max.Z >= blockBB->Min.Z && entityBB->Min.Z <= blockBB->Max.Z) *tz = 0.0f; /* Inlined ZIntersects */
}

void Searcher_Free(void) {
	if (Searcher_StatesCount > SEARCHER_STATES_MIN) {
		Platform_MemFree(Searcher_States);
	}
}